js/ag_reference.js

```diff
@@ -0,0 +1,299 @@
+    this.reference = reference;
+    this.hasPrevious = hasPrevious;
+    this.hasNext = hasNext;
+    
+    //TODO use the data somehow
+}
+AttributeGroupPage.prototype = {
+    constructor: AttributeGroupPage,
+    
+    get tuples () {
+        //TODO
+        notimplemented();
+    },
+    
+    get next() {
+        //TODO same reference with the new after
+        notimplemented();
+    },
+    
+    get previous() {
+        //TODO same reference with the new before
+        notimplemented();
+    }
+};
+
+function AttributeGroupTuple(page, data) {
+    this.page = page;
+    
+    //TODO use the data
+}
+AttributeGroupTuple.prototype = {
+    constructor: AttributeGroupTuple,
+    
+    // TODO take a look at values comment.
+    get isHTML() {
+        if (this._isHTML === undefined) {
+            var value = this.values;
+        }
+        return this._isHTML;
+    },
+
+    get values() {
+        //TODO 
+        // returned value should be the same as current tuple.values 
+        // or we should make this (https://github.com/informatics-isi-edu/ermrestjs/issues/189) change now?
+        // chaise is either way changing it a object with isHTMl and 
+        notimplemented();
+    },
+    
+    get uniqueId() {
+        //TODO
+        notimplemented();
+    },
+    
+    get displayname() {
+        //TODO this is the rowname
+        notimplemented();
+    }
+};
+
+function AttributeGroupColumn(alias, name, displayname, type, comment, sortable, visible) {
+    /**
+     * @private
+     * @type {string}
+     */
+    this._alias = alias;
+    
+    /**
+     * This might include the aggregate functions,
+     * should we seperate those?
+     * @type {string}
+     */
+    this._name = name;
+    
+    /**
+     * TODO not sure if it's needed or not
+     * @type {object}
+     */
+    this.displayname = displayname;
+    
+    /**
+     * Type object
+     * @type {ERMrest.Type}
+     */
+    this.type = type;
+    
+    /**
+     * tooltip
+     * @type {string}
+     */
+    this.comment = comment;
+    
+    /**
+     * [sortable description]
+     * @type {boolean}
+     */
+    this.sortable = sortable; // sort only based on current column, what about sort basesd on other columns?
+    
+    /**
+     * We should have a concept of visible columns, this was the easiest way of implementing it to me.
+     * @type {boolean}
+     */
+    this._visible = visible;
+}
+AttributeGroupColumn.prototype = {
+    constructor: AttributeGroupColumn,
+    
+    // TODO or alias is required? 
+    get name() {
+        if (typeof this._alias === "string" && this._alias.length !== 0) {
+            return this._alias;
+        }
+        return this._name;
+    },
+    
+    formatvalue: function (data, options) {
+        //TODO should be the same as Column.formatvalue
+        notimplemented();
+    },
+    
+    formatPresentation: function (data, options) {
+        //TODO take cares of markdown_pattern and all, is it needed?
```

Comment:
I don't know how the user would define a markdown pattern for one of these columns. We would have to design and annotation for it as well.